### ISSUE TITLE
Generalize `hough_line_peaks` to work on circular, ellipse, and other Hough transforms

### ISSUE THREAD

**Issue by stefanv** [scikit-image/scikit-image#1964]
Currently, `hough_line_peaks` just works on the accumulator from the Hough line transform.  But its non-maximum suppression would be useful for the other Hough transforms too.

**Comment by sciunto** [comment]
That's a great idea. Since we have a new function to write (hough_peaks I guess), we can also the benefit to include a subpixel resolution as an optional feature of this detector (see #2089)

**Comment by sciunto** [comment]
Quick note. What you propose is actually doable for hough_line and hough_circle. Not hough_ellipse. If you look at the API of hough_ellipse, you'll see that it's different. We don't return a Hough space. The reason is that the spirit of the hough ellipse algo is radically different for optimization purpose. For ellipses, it returns accumulators and associated ellipse parameters (5D).

For circles, it's doable. We can imagine that for each radius, we seek the peak position in the (x-y) plane.

I started to refactor hough_line_peak to get a more general function out of it. Then, I'll make a proposition for a new API to serve houghline and houghcircle.

**Comment by sciunto** [comment]
I refactored the code to make things easier to (re)use. I'm waiting #2089  to be merged before starting a new PR with this.

It's fully backward compatible and I checked that it gives the same results as previous implementations in the gallery. In addition, we have a function that could apply to any kind of images to detect peak positions. I suggest that this could go to `features/peaks.py`

```python

from scipy import ndimage as ndi
from skimage import measure

# TODO: check that what I call x and y is consistent with skimage def.
# I never know... :(

def hough_circle_peaks(hspaces, radii, min_xdistance=1, min_ydistance=1, 
                       threshold=None, num_peaks=np.inf, total_num_peaks=np.inf):
    """Return peaks in a circle Hough transform.

    Identifies most prominent circles separated by certain distances
    in a Hough transform. Non-maximum suppression with different sizes is
    applied separately in the first and second dimension
    of the Hough space to identify peaks.

    Parameters
    ----------
    hspaces : (N, M) array
        Hough spaces returned by the `hough_circle` function.
    radii : (M,) array
        Radii corresponding to Hough spaces.
    min_xdistance : int
        Minimum distance separating centers in the x dimension.
    min_ydistance : int
        Minimum distance separating centers in the y dimension.
    threshold : float
        Minimum intensity of peaks in each Hough space. 
        Default is `0.5 * max(hspace)`.
    num_peaks : int
        Maximum number of peaks in each Hough space. When the 
        number of peaks exceeds `num_peaks`, only `num_peaks` 
        coordinates based on peak intensity are considered for the
        corresponding radius.
    total_num_peaks : int
        Maximum number of peaks. When the number of peaks exceeds `num_peaks`,
        return `num_peaks` coordinates based on peak intensity.

    Returns
    -------
    accum, cx, cy, rad : tuple of array
        Peak values in Hough space, x and y center coordinates and radii.

    Examples
    --------
    >>> from skimage import transform as tf 
    >>> from skimage import draw
    >>> img = np.zeros((120, 100), dtype=int)
    >>> radius, x_0, y_0 = (20, 99, 50)
    >>> y, x = draw.circle_perimeter(y_0, x_0, radius)
    >>> img[x, y] = 1
    >>> hspaces = tf.hough_circle(img, radius)
    >>> accum, cx, cy, rad = hough_circle_peaks(hspaces, [radius,])
    """
    
    r = []
    cx = []
    cy = []
    accum = []
    
    for rad, hp in zip(radii, hspaces):
        h_p, x_p, y_p = prominent_peaks(hp, 
                                        min_xdistance=min_xdistance, 
                                        min_ydistance=min_ydistance, 
                                        threshold=threshold, 
                                        num_peaks=num_peaks)
        
        r.extend((rad,)*len(h_p))
        cx.extend(x_p)
        cy.extend(y_p)
        accum.extend(h_p)
        
    r = np.array(r)
    cx = np.array(cx)
    cy = np.array(cy)
    accum = np.array(accum)
    s = np.argsort(accum)
      
    if total_num_peaks != np.inf:
        tnp = total_num_peaks - 1       
        return (accum[s][:tnp:-1], cx[s][:tnp:-1], cy[s][:tnp:-1], r[s][:tnp:-1])
    
    return (accum[s][::-1], cx[s][::-1], cy[s][::-1], r[s][::-1])


#TODO: docstring: assumed to be continuous. What's the meaning?
# Regularly spaced?
def hough_line_peaks(hspace, angles, dists, min_angle=10, min_distance=9,
                     threshold=None, num_peaks=np.inf):
    """Return peaks in a straight line Hough transform.

    Identifies most prominent lines separated by a certain angle and distance
    in a Hough transform. Non-maximum suppression with different sizes is
    applied separately in the first (distances) and second (angles) dimension
    of the Hough space to identify peaks.

    Parameters
    ----------
    hspace : (N, M) array
        Hough space returned by the `hough_line` function.
    angles : (M,) array
        Angles returned by the `hough_line` function. Assumed to be continuous.
        (`angles[-1] - angles[0] == PI`).
    dists : (N, ) array
        Distances returned by the `hough_line` function.
    min_angle : int
        Minimum angle separating lines (maximum filter size for second
        dimension of hough space).
    min_distance : int
        Minimum distance separating lines (maximum filter size for first
        dimension of hough space).
    threshold : float
        Minimum intensity of peaks. Default is `0.5 * max(hspace)`.
    num_peaks : int
        Maximum number of peaks. When the number of peaks exceeds `num_peaks`,
        return `num_peaks` coordinates based on peak intensity.

    Returns
    -------
    accum, angles, dists : tuple of array
        Peak values in Hough space, angles and distances.

    Examples
    --------
    >>> from skimage.transform import hough_line, hough_line_peaks
    >>> from skimage.draw import line
    >>> img = np.zeros((15, 15), dtype=np.bool_)
    >>> rr, cc = line(0, 0, 14, 14)
    >>> img[rr, cc] = 1
    >>> rr, cc = line(0, 14, 14, 0)
    >>> img[cc, rr] = 1
    >>> hspace, angles, dists = hough_line(img)
    >>> hspace, angles, dists = hough_line_peaks(hspace, angles, dists)
    >>> len(angles)
    2

    """    
    h, a, d = prominent_peaks(hspace, min_xdistance=min_angle, 
                          min_ydistance=min_distance, threshold=threshold, 
                          num_peaks=num_peaks)
    return (h, angles[a], dists[d])


#To go to features/peaks.py
def prominent_peaks(image, min_xdistance=1, min_ydistance=1, 
                threshold=None, num_peaks=np.inf):
    """Return peaks with non-maximum suppression.

    Identifies most prominent features separated by certain distances. 
    Non-maximum suppression with different sizes is applied separately 
    in the first and second dimension of the image to identify peaks.

    Parameters
    ----------
    image : ndarray
        Input image.
    min_xdistance : int
        Minimum distance separating features in the x dimension.
    min_ydistance : int
        Minimum distance separating features in the y dimension.
    threshold : float
        Minimum intensity of peaks. Default is `0.5 * max(image)`.
    num_peaks : int
        Maximum number of peaks. When the number of peaks exceeds `num_peaks`,
        return `num_peaks` coordinates based on peak intensity.

    Returns
    -------
    intensity, xcoords, ycoords : tuple of array
        Peak intensity values, x and y indices.
    """

    img = image.copy()
    rows, cols = img.shape

    if threshold is None:
        threshold = 0.5 * np.max(img)

    ycoords_size = 2 * min_ydistance + 1
    xcoords_size = 2 * min_xdistance + 1
    img_max = ndi.maximum_filter1d(img, size=ycoords_size, axis=0,
                                      mode='constant', cval=0)
    img_max = ndi.maximum_filter1d(img_max, size=xcoords_size, axis=1,
                                      mode='constant', cval=0)
    mask = (img == img_max)
    img *= mask
    img_t = img > threshold

    label_img = measure.label(img_t)
    props = measure.regionprops(label_img, img_max)

    # Sort the list of peaks by intensity, not left-right, so larger peaks
    # in Hough space cannot be arbitrarily suppressed by smaller neighbors
    props = sorted(props, key=lambda x: x.max_intensity)[::-1]
    coords = np.array([np.round(p.centroid) for p in props], dtype=int)

    img_peaks = []
    ycoords_peaks = []
    xcoords_peaks = []

    # relative coordinate grid for local neighbourhood suppression
    ycoords_ext, xcoords_ext = np.mgrid[-min_ydistance:min_ydistance + 1,
                                   -min_xdistance:min_xdistance + 1]

    for ycoords_idx, xcoords_idx in coords:
        accum = img_max[ycoords_idx, xcoords_idx]
        if accum > threshold:
            # absolute coordinate grid for local neighbourhood suppression
            ycoords_nh = ycoords_idx + ycoords_ext
            xcoords_nh = xcoords_idx + xcoords_ext

            # no reflection for distance neighbourhood
            ycoords_in = np.logical_and(ycoords_nh > 0, ycoords_nh < rows)
            ycoords_nh = ycoords_nh[ycoords_in]
            xcoords_nh = xcoords_nh[ycoords_in]

            # reflect xcoords and assume xcoords are continuous, 
            # e.g. for angles:
            # (..., 88, 89, -90, -89, ..., 89, -90, -89, ...)
            xcoords_low = xcoords_nh < 0
            ycoords_nh[xcoords_low] = rows - ycoords_nh[xcoords_low]
            xcoords_nh[xcoords_low] += cols
            xcoords_high = xcoords_nh >= cols
            ycoords_nh[xcoords_high] = rows - ycoords_nh[xcoords_high]
            xcoords_nh[xcoords_high] -= cols

            # suppress neighbourhood
            img_max[ycoords_nh, xcoords_nh] = 0

            # add current feature to peaks
            img_peaks.append(accum)
            ycoords_peaks.append(ycoords_idx)
            xcoords_peaks.append(xcoords_idx)

    img_peaks = np.array(img_peaks)
    ycoords_peaks = np.array(ycoords_peaks)
    xcoords_peaks = np.array(xcoords_peaks)

    if num_peaks < len(img_peaks):
        idx_maxsort = np.argsort(img_peaks)[::-1][:num_peaks]
        img_peaks = img_peaks[idx_maxsort]
        ycoords_peaks = ycoords_peaks[idx_maxsort]
        xcoords_peaks = xcoords_peaks[idx_maxsort]

    return img_peaks, xcoords_peaks, ycoords_peaks

```
